//  Copyright 2018 U.C. Berkeley RISE Lab
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

syntax = "proto2";

message Function {
  required string name = 1;
  required bytes body = 2;
}

enum SerializerType {
  DEFAULT = 0;
  STRING = 1;
  NUMPY = 2;
}

message Value {
  required bytes body = 1;
  optional SerializerType type = 2;
}

message FunctionCall {
  required string name = 1;
  required uint32 request_id = 2;
  repeated Value args = 3;
  optional string resp_id = 4;
}

message FunctionList {
  repeated string names = 1;
}

message Dag {
  message DagPair {
    required string source = 1;
    required string sink = 2;
  }

  required string name = 1;
  repeated string functions = 2;
  repeated DagPair connections = 3;
}

message DagCall {
  message DagStageArgs {
    required string name = 1;
    repeated Value args = 2;
  }

  required string name = 1;
  repeated DagStageArgs args = 2;
}
